### PR TITLE
Correctly await outputs containing resources

### LINKED ISSUE #9008: pulumi up panics with Go sdk - panic: fatal: An assertion has failed
## Issue details

Creating an EKS cluster using pulumi Go sdk results in panic.

```
Resources:
    + 26 to create
    2 unchanged

Do you want to perform this update? yes
Updating (eks-test):
     Type                 Name             Status         Info
     pulumi:pulumi:Stack  eks_go-eks-test  **failed**     1 error; 23 messages

Diagnostics:
  pulumi:pulumi:Stack (eks_go-eks-test):
    error: an unhandled error occurred: program exited with non-zero exit code: 1

    panic: fatal: An assertion has failed
    goroutine 98 [running]:
    github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
        /home/vikas/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.24.1/go/common/util/contract/failfast.go:23
    github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Assert(...)
        /home/vikas/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.24.1/go/common/util/contract/assert.go:26
    github.com/pulumi/pulumi/sdk/v3/go/pulumi.marshalInputImpl({0x14e7120, 0xc0004ec140}, {0x1781850, 0x14e7120}, 0x1, 0x0)
        /home/vikas/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.24.1/go/pulumi/rpc.go:353 +0x263b
    github.com/pulumi/pulumi/sdk/v3/go/pulumi.marshalInputImpl({0x1468920, 0xc0005875e0}, {0x1781850, 0x14e7120}, 0x1, 0x0)
        /home/vikas/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.24.1/go/pulumi/rpc.go:273 +0xc5d
    github.com/pulumi/pulumi/sdk/v3/go/pulumi.marshalInput(...)
        /home/vikas/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.24.1/go/pulumi/rpc.go:217
    github.com/pulumi/pulumi/sdk/v3/go/pulumi.marshalInputs.func1({0x10cecaf, 0xc}, {0x1468920, 0xc0005875e0}, {0x1781850, 0x14e7120})
        /home/vikas/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.24.1/go/pulumi/rpc.go:134 +0x8b
    github.com/pulumi/pulumi/sdk/v3/go/pulumi.marshalInputs({0x172b280, 0xc0000d8000})
        /home/vikas/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.24.1/go/pulumi/rpc.go:187 +0x878
    github.com/pulumi/pulumi/sdk/v3/go/pulumi.(*Context).prepareResourceInputs(0xc0004ec000, {0x177a9a8, 0xc00014a200}, {0x172b280, 0xc0000d8000}, {0x1591684, 0x11}, 0xc00014d200, 0xc0004ac300, 0x1)
        /home/vikas/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.24.1/go/pulumi/context.go:1213 +0x1a5
    github.com/pulumi/pulumi/sdk/v3/go/pulumi.(*Context).registerResource.func1()
        /home/vikas/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.24.1/go/pulumi/context.go:794 +0x27f
    created by github.com/pulumi/pulumi/sdk/v3/go/pulumi.(*Context).registerResource
        /home/vikas/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.24.1/go/pulumi/context.go:781 +0x7a8
    exit status 2

Resources:
    2 unchanged

Duration: 3s
```

### Steps to reproduce

Relevant code:

```
func NewEksCluster(ctx *pulumi.Context, cfg *config.Config) (*eks.Cluster, error) {
	awsRegion := cfg.Require("aws_region")
	awsProviderName := cfg.Require("aws_provider_name")
	awsProfile := cfg.Require("aws_profile")

	awsProvider, err := aws.NewProvider(ctx, awsProviderName, &aws.ProviderArgs{
		Region:  pulumi.StringPtr(awsRegion),
		Profile: pulumi.StringPtr(awsProfile),
	})
	if err != nil {
		return nil, fmt.Errorf("Failed to create aws provider: %w", err)
	}

	instanceRoleName := cfg.Require("aws_instance_role")
	serviceRoleName := cfg.Require("aws_service_role")

	instanceRole, err := getRole(ctx, instanceRoleName, awsProvider)
	if err != nil {
		return nil, err
	}
	serviceRole, err := getRole(ctx, serviceRoleName, awsProvider)
	if err != nil {
		return nil, err
	}

	clusterName := cfg.Require("clusterName")
	var subnetIds []string
	cfg.RequireObject("subnetIds", &subnetIds)

	cluster, err := eks.NewCluster(ctx, clusterName, &eks.ClusterArgs{
		Name:                  pulumi.StringPtr(clusterName),
		Version:               pulumi.StringPtr(cfg.Require("masterVersion")),
		MinSize:               pulumi.IntPtr(cfg.RequireInt("minNodes")),
		MaxSize:               pulumi.IntPtr(cfg.RequireInt("maxNodes")),
		InstanceType:          pulumi.StringPtr(cfg.Require("instanceType")),
		NodeAmiId:             pulumi.StringPtr(cfg.Get("nodeAmiId")),
		VpcId:                 pulumi.StringPtr(cfg.Require("vpcId")),
		SubnetIds:             pulumi.ToStringArray(subnetIds),
		InstanceRole:          instanceRole.ToRoleOutput(),
		ServiceRole:           serviceRole.ToRoleOutput(),
		InstanceProfileName:   pulumi.StringPtr(cfg.Require("instanceProfileName")),
		EndpointPublicAccess:  pulumi.BoolPtr(cfg.RequireBool("endpointPublicAccess")),
		EndpointPrivateAccess: pulumi.BoolPtr(cfg.RequireBool("endpointPrivateAccess")),
		UseDefaultVpcCni:      pulumi.BoolPtr(true),
		RoleMappings:          eks.RoleMappingArray{},
		ProviderCredentialOpts: &eks.KubeconfigOptionsArgs{
			ProfileName: pulumi.StringPtr(awsProfile),
		},
	},
		pulumi.Provider(awsProvider),
	)

	if err != nil {
		return nil, fmt.Errorf("Failed to create eks cluster: %w", err)
	}

	if err := addNodeGroups(ctx, cluster, cfg, awsProvider); err != nil {
		return cluster, fmt.Errorf("Failed to create nodegroups: %w", err)
	}

	return cluster, nil

}
```

There seem to have been a [fix](https://github.com/pulumi/pulumi/pull/6962) that addresses a [similar issue](https://github.com/pulumi/pulumi/issues/6934).
My suspicion is that it has to do with the resource references to the IAM roles.

## Environment details

```
✗ pulumi version
v3.24.1

✗ pulumi plugin ls
NAME        KIND      VERSION  SIZE    INSTALLED  LAST USED
aws         resource  4.37.4   350 MB  n/a        14 minutes ago
aws         resource  4.34.0   358 MB  n/a        14 minutes ago
aws         resource  4.27.2   344 MB  n/a        14 minutes ago
aws         resource  4.27.0   344 MB  n/a        14 minutes ago
docker      resource  3.1.0    40 MB   n/a        14 minutes ago
eks         resource  0.37.1   174 MB  n/a        11 minutes ago
eks         resource  0.35.0   174 MB  n/a        14 minutes ago
kubernetes  resource  3.16.0   77 MB   n/a        11 minutes ago
kubernetes  resource  3.15.2   77 MB   n/a        14 minutes ago
kubernetes  resource  3.14.0   77 MB   n/a        14 minutes ago
kubernetes  resource  3.10.0   77 MB   n/a        14 minutes ago
kubernetes  resource  3.9.0    77 MB   n/a        14 minutes ago

```

### PR DESCRIPTION
**Pull request by iwahbe**
# Description

Fixes #9008 

## Checklist

- [X] I have added tests that prove my fix is effective or that my feature works

- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by pgavlin** [comment]
Working on adding some comments to this.

File: sdk/go/pulumi/types_test.go [hunk 1]
```diff
@@ -339,6 +339,11 @@ func TestToOutputAnyDeps(t *testing.T) {
 	}()
 
 	res := &ResourceState{}
+	urnOut := URNOutput{newOutputState(nil, reflect.TypeOf(URN("")), res)}
+	go func() {
```
Comment (Frassle) [inline sdk/go/pulumi/types_test.go:343]: Why's this in a goroutine?

File: sdk/go/pulumi/types_test.go [hunk 2]
```diff
@@ -352,7 +357,7 @@ func TestToOutputAnyDeps(t *testing.T) {
 	v, known, secret, deps, err := await(out)
 	assert.True(t, known)
 	assert.False(t, secret)
-	assert.ElementsMatch(t, []Resource{stringDep1, stringDep2, intDep1, intDep2, boolDep1, boolDep2, res}, deps)
+	assert.ElementsMatch(t, []Resource{stringDep1, stringDep2, intDep1, intDep2, boolDep1, boolDep2}, deps)
```
Comment (Frassle) [inline sdk/go/pulumi/types_test.go:360]: Why doesn't this depend on res anymore?

File: sdk/go/pulumi/types.go [hunk 3]
```diff
@@ -717,7 +735,7 @@ func awaitInputs(ctx context.Context, v, resolved reflect.Value) (bool, bool, []
 			valueType = v.Type()
 		} else {
 			// Handle pointer inputs.
-			if v.Kind() == reflect.Ptr {
+			if v.Kind() == reflect.Ptr && !(v.CanInterface() && v.Type().Implements(resourceType)) {
```
Comment (Frassle) [inline sdk/go/pulumi/types.go:738]: Merge master, I've just changed this same bit of code to handle the case of v being a pointer but resolved being interface{}

File: sdk/go/pulumi/types.go [hunk 4]
```diff
@@ -728,6 +746,11 @@ func awaitInputs(ctx context.Context, v, resolved reflect.Value) (bool, bool, []
 
 	contract.Assertf(valueType.AssignableTo(resolved.Type()), "%s not assignable to %s", valueType.String(), resolved.Type().String())
 
+	if v.CanInterface() && v.Type().Implements(resourceType) {
```
Comment (Frassle) [inline sdk/go/pulumi/types.go:749]: Do we need the CanInterface check here? Isn't that covered by Implements?

File: sdk/go/pulumi/types.go [hunk 5]
```diff
@@ -728,6 +746,11 @@ func awaitInputs(ctx context.Context, v, resolved reflect.Value) (bool, bool, []
...
+		resolved.Set(v)
+		return true, false, nil, nil
```
Comment (Frassle) [inline sdk/go/pulumi/types.go:751]: Shouldn't deps here include v.urn?

File: sdk/go/pulumi/types_test.go (hunk 1 above)
Comment (AaronFriel) [reply sdk/go/pulumi/types_test.go:343]: @iwahbe this is my question as well - does `urnOut.resolve()` have side effects that occur in the same thread?

File: sdk/go/pulumi/types_test.go (hunk 2 above)
Comment (pgavlin) [reply sdk/go/pulumi/types_test.go:360]: The dependency on `res` is inconsistent with the behavior of `All` in other languages, which do not include resource instances encountered during the crawl in the list of deps for the output.

File: sdk/go/pulumi/types_test.go (hunk 1 above)
Comment (pgavlin) [reply sdk/go/pulumi/types_test.go:343]: Putting this in a goroutine probabilistically tests that the code is able to properly await unresolved outputs.

File: sdk/go/pulumi/types.go (hunk 4 above)
Comment (pgavlin) [reply sdk/go/pulumi/types.go:749]: It is not, no--`CanInterface` just indicates whether or not the `Interface()` method of `reflect.Value` will panic.

Comment (pgavlin) [reply sdk/go/pulumi/types.go:749]: Ah, though we don't need the check b/c we're not _calling_ `Interface()`.

File: sdk/go/pulumi/types.go (hunk 5 above)
Comment (pgavlin) [reply sdk/go/pulumi/types.go:751]: No--including the URNs of resource instances encountered is inconsistent with the other SDKs.

**Comment by AaronFriel** [comment]
@pgavlin or @iwahbe is this in a merge-able state?

**Comment by iwahbe** [comment]
I don't know. @pgavlin?

**Comment by pgavlin** [comment]
This is finally ready to go.

**Review by AaronFriel — APPROVED** [review]

**Pull request merged**